src/Stratis.Bitcoin/Base/ConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,434 @@
+        /// by the peer when a new block is validated (unless in IBD).
+        /// <para>
+        /// When we receive "headers" message from the peer, we can adjust our knowledge
+        /// of the peer's view of the chain. We update its pending tip, which represents
+        /// the tip of the best chain we think the peer has.
+        /// </para>
+        /// <para>
+        /// If we receive a valid header from peer which work is higher than the work
+        /// of our best chain's tip, we update our view of the best chain to that tip.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessHeadersAsync(INetworkPeer peer, HeadersPayload headersPayload)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(headersPayload), headersPayload);
+
+            List<BlockHeader> headers = headersPayload.Headers;
+
+            if (headers.Count == 0)
+            {
+                this.logger.LogTrace("Headers payload with no headers was received. Assuming we're synced with the peer.");
+                this.logger.LogTrace("(-)[NO_HEADERS]");
+                return;
+            }
+
+            // Check headers for consecutiveness.
+            for (int i = 1; i < headers.Count; ++i)
+            {
+                if (headers[i].HashPrevBlock != headers[i - 1].GetHash())
+                {
+                    this.peerBanning.BanAndDisconnectPeer(peer.PeerEndPoint, this.connectionManager.ConnectionSettings.BanTimeSeconds, "Peer presented nonconsecutive headers.");
+
+                    this.logger.LogTrace("(-)[PEER_BANNED]");
+                    return;
+                }
+            }
+
+            using (await this.asyncLock.LockAsync().ConfigureAwait(false))
+            {
+                if (this.cachedHeaders.Count > CacheSyncHeadersThreshold)
+                {
+                    // Ignore this message because cache is full.
+                    return;
+                }
+
+                // If queue is not empty- add to queue instead of calling CM.
+                if (this.cachedHeaders.Count != 0)
+                {
+                    this.cachedHeaders.AddRange(headers);
+
+                    this.logger.LogTrace("{0} headers were added to cache.", headers.Count);
+                }
+                else
+                {
+                    ConnectNewHeadersResult result = await this.PresentHeadersLockedAsync(headers).ConfigureAwait(false);
+
+                    if (result.Consumed == null)
+                    {
+                        this.cachedHeaders.AddRange(headers);
+                        this.logger.LogTrace("None of {0} headers were consumed, all were added to cache.", headers.Count);
+                    }
+                    else
+                    {
+                        this.ExpectedPeerTip = result.Consumed;
+
+                        if (result.Consumed.Header != headers.Last())
```

Comment:
Nested `if'`s, I feel this can be written better.

Reply:
Fixed